### ISSUE TITLE
Allow many:many relationships in MappedEntityDomainManager

### ISSUE THREAD

**Issue by migig** [Azure/azure-mobile-apps-net-server#159]
**TL;DR
My pull request adds a very simple workaround for the lack of many:many relationships.**

A _very_ common complaint is that the SDK doesn't support 1:many and many:many relationships. I've read in multiple locations (blogs, StackOverflow) that this will be added at some point. However that may take ages, as it's a large feature for you to develop and test.

For your customers who are developing non-trivial apps (not a "TodoApp"), we need that functionality now, and we can't wait. So we're forced to roll our own, because the [`MappedEntityDomainManager`](https://github.com/Azure/azure-mobile-apps-net-server/blob/master/src/Microsoft.Azure.Mobile.Server.Entity/MappedEntityDomainManager.cs) isn't capable enough for a complex domain.

After days and days of hacking, I've discovered that the only real way to solve this problem, without tossing table controllers out the window altogether, is to subclass `MappedEntityDomainManager` and heavily modify all calls where mapping is done. But it calls into internal and private methods, so I need to duplicate them too, and duplicate stuff they call into (in other classes), and so on, and so on! It's a mess, though it works. But I now need to maintain infrastucture code which isn't mine, and I don't even know where to begin in testing it... :cry:

_I have found a workaround which is non-breaking and easy to apply._

As I explained in [another issue](http://github.com/Azure/azure-mobile-apps-net-server/issues/157), the basic problem is that a complex entity has navigation properties which cannot be easily/properly mapped by AutoMapper. And as I explained, even if possible, it's a very bad idea. The `MappedEntityDomainManager` just assumes that everyone will map this way, and expects those reverse mappings to be configured. If they aren't, it throws.

A better approach is to provide hooks for us to use. The class should add virtual `MapDtoToEntity()` methods, which use AutoMapper exactly as it is done now - but if we need to perform custom DTO-to-entity mapping, then we can override it and call our mapping services instead. This is a non-breaking change and would not affect existing code.

I've [added a sample PR here](http://github.com/Azure/azure-mobile-apps-net-server/pull/158), I expect it should pass existing tests as it is functionally the same as the current code.

If you add this code, anyone who complains about lack of many:many support can be told to shut up already! :smile:

Assuming it passes your tests, and it should, it is a VERY simple fix... Please adopt it and push to nuget!

**Comment by migig** [comment]
While you're at it, it may be a good idea to also think ahead and add a pair of methods with the following signature:

    protected virtual TData MapEntityToDto(TModel entity) {
        return Mapper.Map<TModel, TData>(entity);
    }

    protected virtual TData MapEntityToDto(TModel entity, TData dto) {
        return Mapper.Map<TModel, TData>(entity, dto);
    }

...And update all entity-to-DTO mappings to use these new methods.

**Comment by adrianhall** [comment]
Adding transparency, this server won't be seeing relationships as a default manner, so closing this issue.
